### PR TITLE
[PERF] Release some index spec info off the main thread

### PR DESCRIPTION
**Pull request by ashtul**
For a large index, releasing all internals might be a major task that costumes much time.

In this PR, all data that is not connected to any global data or does not require the GIL is free on a thread. The drop in consumed time is significant and would prevent a watchdog from killing the shard.

Related to MOD-2364

### CODE REVIEW

**Comment by MeirShpilraien** [comment]
> In this PR, all data that is not connected to any global data or does not require the GIL is free on a thread. The drop in consumed time is significant and would prevent a watchdog from killing the shard.

Please provide the full list? What is going to be freed on the background thread?

File: src/spec.c [hunk 1]
```diff
@@ -1069,21 +1052,69 @@ void IndexSpec_FreeInternals(IndexSpec *spec) {
...
+    spec->sortables = NULL;
   }
 
-  if (spec->keysDict) {
-    dictRelease(spec->keysDict);
-  }
+  rm_free(spec);
+}
 
+void IndexSpec_FreeInternals(IndexSpec *spec) {
```
Comment (MeirShpilraien) [inline src/spec.c:1066]: Please explain in a comment what is freed here and what is going to be freed on a background thread, Also maybe call it unlink?

Comment (MeirShpilraien) [reply src/spec.c:1066]: I see you have it on function body so forget it, maybe just add a comment explains that this is done on the main thread with the GIL locked and then the rest of the free operation goes to background thread.

File: src/spec.c [hunk 2]
```diff
@@ -1069,21 +1052,69 @@ void IndexSpec_FreeInternals(IndexSpec *spec) {
...
+  // Free stopwords list (might use global pointer to default list)
+  if (spec->stopwords) {
+    StopWordList_Unref(spec->stopwords);
+    spec->stopwords = NULL;
+  }
+  // Free data complex data structures on a different thread
+  if (getenv("RS_GLOBAL_DTORS")) {  // used only with sanitizer or valgrind
+    IndexSpec_FreeData(spec);
+  } else {
+    if (!cleanPool) {
```
Comment (MeirShpilraien) [inline src/spec.c:1113]: I think we can initialise it on start

File: tests/pytests/test.py [hunk 3]
```diff
@@ -3497,3 +3497,39 @@ def test_empty_field_name(env):
     env.expect('FT.CREATE', 'idx', 'SCHEMA', '', 'TEXT').ok()
     conn.execute_command('hset', 'doc1', '', 'foo')
     env.expect('FT.SEARCH', 'idx', 'foo').equal([1L, 'doc1', ['', 'foo']])
+
+
+def test_clear_fields(env):
```
Comment (MeirShpilraien) [inline tests/pytests/test.py:3502]: How long does it take? Maybe run it on nightly? Or maybe add it to the performance tests?

File: src/spec.c [hunk 4]
```diff
@@ -1069,21 +1052,69 @@ void IndexSpec_FreeInternals(IndexSpec *spec) {
     }
     rm_free(spec->fields);
   }
-  if (spec->uniqueId) {
-    // If uniqueid is 0, it means the index was not initialized
-    // and is being freed now during an error.
-    IndexSpec_ClearAliases(spec);
+  // ahhhm
```
Comment (MeirShpilraien) [inline src/spec.c:1055]: ?

**Review by MeirShpilraien — COMMENTED** [review]
:+1: Small comments

**Review by MeirShpilraien — APPROVED** [review]

File: tests/pytests/test.py (hunk 3 above)
Comment (ashtul) [reply tests/pytests/test.py]: Added module param and made a test that runs twice with and without freeing resources on the main thread.
The test checks the freeing on a 2nd thread is more than 10 times faster.

**Review by MeirShpilraien — COMMENTED** [review]
One comment

File: src/config.h [hunk 5]
```diff
@@ -186,7 +186,7 @@ sds RSConfig_GetInfoString(const RSConfig *config);
     .maxSearchResults = SEARCH_REQUEST_RESULTS_MAX, .maxAggregateResults = -1,                    \
     .minUnionIterHeap = 20, .numericCompress = false, .numericTreeMaxDepthRange = 0,              \
     .printProfileClock = 1, .invertedIndexRawDocidEncoding = false,                               \
-    .forkGCCleanNumericEmptyNodes = true, .freeResources = false,                                 \
+    .forkGCCleanNumericEmptyNodes = true, .freeResourcesThread = false,                           \
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/config.h]: @MeirShpilraien changed

**Review by MeirShpilraien — APPROVED** [review]

**Pull request merged**